Fix hang when retries fail.

ccstolley opened the pull request:
The getter can hang when we manage to get past the initial request
for the content length, but all subsequent chunk requests fail.
When this happens, the sending (`worker()`) side never closes the
`g.readCh` channel, and thus the reading side blocks forever on select.

This change adds a `workerAborted` channel to signal to Read() callers that
something went wrong while trying to fetch all the chunks. As multiple worker
goroutines may encounter problems, we wrap the `close(workerAborted)` call
in a `sync.Once` to avoid `panic()`ing by closing a closed channel.

Tests are included to both verify the broken behavior prior to the
change (the test hangs without this fix) and verify the happy path
still works as expected.

cc @carlosmn

Commit:
Fix hang when retries fail.

The getter can hang when we manage to get past the initial request
for the content length, but all subsequent chunck requests fail.
When this happens, the sending (worker()) side never closes the
readCh channel, and thus the reading side blocks forever on select.

This change closes readCh if retryGetChunk() fails, thus notifying
the reading side that we gave up and to abort with an error.

Tests are included to both verify the broken behavior prior to the
change (the test hangs without this fix) and verify the happy path
still works as expected.

carlosmn commented on getter.go:
```diff
@@ -152,26 +152,30 @@ func (g *getter) initChunks() {
 	for c := range g.getCh {
-		g.retryGetChunk(c)
+		if err := g.retryGetChunk(c); err != nil {
+			close(g.readCh)
```
If multiple workers have issues, this will make each of them close the channel potentially leading to a panic.

ccstolley replied:
Great catch, thanks. I'll have to think about that a bit more.

ccstolley replied:
How about wrapping it in a `sync.Once` like so:
```suggestion
			g.clOnce.Do(func() { close(g.readCh) })
```

carlosmn replied:
It's not super clear, but I think that would work in lieu of reorganising the flow of channels.

carlosmn commented on getter.go:
```diff
@@ -152,26 +152,30 @@ func (g *getter) initChunks() {
+		return nil
 	default:
 		g.err = err
+		return err
```
Could we rely on reading if `g.err` gets set instead of closing in the reader? Generally you'd only want to close a channel if you're the writer.

ccstolley replied:
I had considered that, but I think checking `g.err` would require taking a lock (though maybe that's not so bad). For what its worth, I understood the `worker()` goroutine to be the [writer](https://github.com/github/s3gof3r/blob/master/getter.go#L205) for the `g.readCh` channel. But the panic issue you raised above may rule out that approach anyway.

Commit:
Make abort condition clearer.

Introduce a new channel to enable signalling to the Read() caller that
the workers have encountered an error (despite retries), and to give up.

Also add a sync.Once to avoid closing the workerAborted channel more
than once.

ccstolley commented:
cc: @github/git-storage and @znull for possible interest.

mhagger commented:
I was looking at this package out of curiousity and noticed that [`checkClose()`](https://github.com/ccstolley/s3gof3r/blob/ccstolley-fix-hang-failed-retries/util.go#L80-L88) never reports errors from `c.Close()`. It only writes the close error into local variable `err`, but that doesn't have any effect at the caller. In fact, I'm surprised the compiler doesn't warn about the ineffectual assignment.

rhettg commented on getter.go:
```diff
@@ -153,6 +158,11 @@ func (g *getter) initChunks() {
 		g.retryGetChunk(c)
+		if g.err != nil {
+			// tell Read() caller that 1 or more chunks can't be read; abort
+			g.abortOnce.Do(func() { close(g.workerAborted) })
```
Oh, that's nice. I just did something similar in a PR [here](https://github.com/github/spokesd/pull/943/files#diff-595c6f3ea9cb61a6a8813c0d7665c8a1f71c1b1de6235eec6fa5af2abb33537eR598-R603). This looks safer

rhettg reviewed (approved):
Pretty sure this makes sense. 

I don't love the idea that any worker could encounter `g.err != nil` and cause the abort, even if it was a different chunk that caused it, but I can't see how it actually matters.

mhagger commented:
Doesn't `g.cond.L.Unlock()` have to be called before [this `return`](https://github.com/ccstolley/s3gof3r/blob/8f0190f23fb61f7d0112a5dc1f8d6f4dfa1e42ef/getter.go#L221)? (E.g. `defer` the `Unlock()` right after the `Lock()`.) It seems like a possible deadlock to me otherwise, if another goroutine is about to try to acquire that lock. Meanwhile, `g.closed` is written without holding a lock, so that also constitutes a data race.

Reading from and writing to `g.err` also seems to happen from multiple goroutines without locking, so that seems to be another data race.

mhagger commented:
This PR is probably fine as is but it occurred to me that the modern ways to kill a bunch of worker goroutine would be one of the following:

* Pass them a `ctx` and cancel the context when an error returns. This could also be easily made to kill any pending web requests whose results won't be needed.
* Use an `errgroup.Group` (probably the `WithContext()` variant).

Either of these would require taking on a new dependency (e.g., the `context` package was only added in Go 1.7, whereas this package claims to require only Go 1.5). Dunno if that's easier overall or would make the code clearer and easier to audit.
